PR to develop from Feature/register-products

mankicho opened the pull request.

mankicho committed:
change application.properties to application.yml

mankicho committed:
Merge branch 'develop' of https://github.com/study-spring/spring-market

mankicho committed:
Add MapperScan

mankicho committed:
Add datasource config

mankicho committed:
Add function to find products

mankicho committed:
Merge branch 'feature/get_products' of https://github.com/mankicho/spring-market

mankicho committed:
rename product -> productVO

mankicho committed:
add find products by using lat,lng coordinates
use spatial query of MySQL

mankicho committed:
add validation dependency

mankicho committed:
apply validation to findProductDTO

mankicho committed:
add request, response object

mankicho committed:
add function to register product

mankicho committed:
rename FindProductResponse to FindProductResponseDTO

Leewongi0731 commented:
범근이는 개발 안함??

Leewongi0731 commented on src/main/java/project/springmarket/controller/ProductController.java:
```diff
@@ -0,0 +1,52 @@
+public class ProductController {
+    private final ProductService productService;
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
```
@ControllerAdvice 걸어서 공통 ExceptionHandler하는 class로 분리해서 처리하는게 좋을 듯

mankicho replied:
옥희

Leewongi0731 commented on src/main/java/project/springmarket/controller/ProductController.java:
```diff
@@ -0,0 +1,52 @@
+        return productService.findProducts(addressNo);
+    }
+
+    @PostMapping("/address/")
```
맨 뒤 /는  빼는게 관행임

mankicho replied:
넹

Leewongi0731 commented on src/main/java/project/springmarket/model/product/ProductStatus.java:
```diff
@@ -0,0 +1,9 @@
+package project.springmarket.model.product;
+
+public enum ProductStatus {
+    판매, 예약, 판매완료;
```
한글보다 영문으로 바꾸는게 좋을 듯..

mankicho replied:
넵

Leewongi0731 commented on src/main/java/project/springmarket/repository/ProductRepository.java:
```diff
@@ -0,0 +1,38 @@
+public class ProductRepository {
+    private final ProductMapper productMapper;
+
+    public FindProductResponseDTO findProducts(int addressNo) {
```
Response는 사용자의 요청에 따른 응답 객체이기 때문에, Repository level에서 만들거나 핸들링하는 것보다, 사용자의 요청을 처리하는 controller level에서 처리하는게 좋아보임. 
 Controller layer : 사용자의 요청/응답 처리 (Request model, Response model)
 Service layer : 비즈니스 로직 처리, 트랜젝션 처리
 Repository layer : DB query 호출

mankicho replied:
오호라 좋은꿀팁 ㅅ

Leewongi0731 commented on src/main/java/project/springmarket/service/ProductService.java:
```diff
@@ -0,0 +1,31 @@
+        this.productRepository = productRepository;
+    }
+
+    public FindProductResponseDTO findProducts(int addressNo) {
```
여기도, Service에서 ResponseDTO 내리지말고, 일반 VO, DTO을 내리고 Controller에서 return할 때 ResponseDTO로 반환하는게 좋아보임

Leewongi0731 commented on src/main/resources/mybatis/mapper/product.xml:
```diff
@@ -0,0 +1,46 @@
+    </resultMap>
+
+    <select id="findProducts" resultMap="products">
+        select *
```
*로 하지말고 가져올 컬럼들 명시 다 하셈
나중에 model 변경되거나 할 때 헷갈림

mankicho replied:
냅

Leewongi0731 commented on src/main/java/project/springmarket/model/product/request/FindProductDTO.java:
```diff
@@ -0,0 +1,21 @@
+
+@AllArgsConstructor
+@Getter
+public class FindProductDTO {
```
product/request/ 에 넣고 쓸거면, FindProductRequestDTO로 변경하는게 좋아보임.

mankicho committed:
rename FindProductDTO to FindProductRequestDTO

mankicho committed:
change status to english

mankicho committed:
change return type
service layer - findProductsUsingPosition : FindProductResponseDTO ->
List<Products>
service layer - findProducts : FindProductResponseDTO -> List<Products>
repository layer - findProductsUsingPosition : FindProductResponseDTO ->
List<Products>
repository layer - findProducts : FindProductResponseDTO ->
List<Products>

mankicho committed:
add advice to handle ValidException

Leewongi0731 commented on src/main/java/project/springmarket/controller/ProductController.java:
```diff
@@ -27,22 +28,28 @@
     public FindProductResponseDTO findProducts(@PathVariable("addressNo") int addressNo) {
-        return productService.findProducts(addressNo);
+        List<ProductVO> products = productService.findProducts(addressNo);
+        return FindProductResponseDTO
```
builder 보다 FindProductResponseDTO에 ProductVO을 인수로 받는 생성자를 만드는게 좋을 듯, 생성자 메소드 이름은 fromProductVO으로 하면 될 듯.

Leewongi0731 commented on src/main/java/project/springmarket/service/ProductService.java:
```diff
@@ -0,0 +1,33 @@
+    private final ProductRepository productRepository;
+
+    @Autowired
+    public ProductService(ProductRepository productRepository) {
```
왜 이렇게 바꾼거야??  RequiredArgsConstructor로 주입해줬었잖아.

mankicho replied:
이거뭐지.. 뭐 실수한듯

Leewongi0731 commented on src/main/java/project/springmarket/controller/ProductController.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @PostMapping("/register")
+    public boolean registerProduct(@RequestBody @Valid ProductRegisterRequestDTO productRegisterRequestDTO) {
```
ProductRegisterRequestDTO 일단, DB등롱이니깐 그냥 insert로 하면 될 것 같고 (ProductRequestDTO) 근데, 이것도 그냥  ProductVO로 받아서 처리하는게 나아보이는데, productNo는 0으로 처리해서... 굳이 RequestDTO가 필요할 까 함

mankicho replied:
넵 수정했습니다

Leewongi0731 commented on src/main/java/project/springmarket/controller/ProductController.java:
```diff
@@ -0,0 +1,47 @@
+                .build();
+    }
+
+    @PostMapping("/register")
```
register 보다 insert를 많이 쓰고, 단순 products insert면 POST /products 면 충분할 듯.  GET/POST/DELETE자체로 의미가 있는 거니깐

mankicho replied:
ㅇㅋㅇㅋ

mankicho committed:
use requiredArgConstructor to DI

mankicho committed:
change builder to constructor

mankicho committed:
change parameter
registerProduct : FindProductRequestDTO -> ProductVO

mankicho committed:
add validation to product model

mankicho committed:
change url /products/register -> /products

mankicho committed:
change parameter to ProductVO

mankicho committed:
rename column category -> categoryNo

mankicho committed:
Merge branch 'develop' into feature/register_products
